okhttp/src/main/java/com/backblaze/b2/client/okHttpClient/B2OkHttpClientImpl.java

```diff
@@ -0,0 +1,317 @@
+        Headers reqHeaders = makeOkHeaders(headersOrNull);
+        if( reqHeaders != null ){
+            builder.headers(reqHeaders);
+        }
+        Request request = builder.build();
+        try (Response response = okHttpClient.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                throw extractExceptionFromErrorResponse(response, response.body().string());
+            } else {
+                if( response.code() == 200){
+                    Headers responseHeaders = response.headers();
+                    return makeB2Headers(responseHeaders);
+                } else {
+                    throw B2Exception.create(null, response.code(), null, "");
+                }
+            }
+        } catch (IOException e) {
+            throw translateToB2Exception(e, url);
+        }
+    }
+
+    @Override
+    public void close() {
+        // todo: Closing response body streams is handled elsewhere by OkHttp. Do we need anything here?
+        // clientFactory.close();
+    }
+
+    private String postJsonAndReturnString(String url,
+                                           B2Headers headersOrNull,
+                                           Object request) throws B2Exception {
+        try {
+            B2Json bzJson = B2Json.get();
+            String requestJson = bzJson.toJson(request);
+            byte[] requestBytes = getUtf8Bytes(requestJson);
+            return postAndReturnString(url, headersOrNull,new ByteArrayInputStream(requestBytes), requestJson.length());
+        } catch (B2JsonException e) {
+            //log.warn("Unable to serialize " + request.getClass() + " using B2Json, was passed in request for " + url, ex);
+            throw new B2LocalException("parsing_failed", "B2Json.toJson(" + request.getClass() + ") failed: " + e.getMessage(), e);
+        }
+
+    }
+
+    private byte[] readFully(InputStream inputStream, int contentLength) throws IOException{
+        int numRead = 0;
+        int offset = 0;
+        int length = contentLength;
+        byte[] bytes = new byte[contentLength];
+        while( numRead < contentLength ){
+            int n = inputStream.read(bytes, offset, length);
+            if( n > -1 ){
+                numRead += n;
+                offset += n;
+                length = contentLength - numRead;
+            }
+        }
+        return bytes;
+    }
+
+    /**
+     * POSTs to a web service that returns content, and returns the content
+     * as a single string.
+     *
+     * @param url the url to post to
+     * @param headersOrNull the headers, if any.
+     *
+     * @return the body of the response.
+     * @throws B2Exception if there's any trouble
+     */
+    private  String postAndReturnString( String url,  B2Headers headersOrNull,  InputStream inputStream, long contentLength)
+            throws B2Exception {
+        byte[] bytes;
+        try {
+            bytes = readFully(inputStream, (int)contentLength);
```

Comment:
maybe use B2Preconditions to make sure that contentLength is small enough to not lose data with this cast?

Reply:
What should I do if it is too large?

Reply:
As discussed, I now use a temporary file when there are too many bytes for a byte[]